src/main.rs

```diff
@@ -69,29 +73,62 @@ impl Client {
     // Method should return an error when a connection already exists.
     // The client should send a handshake to the server.
     fn open(&mut self, addr: &str, server: Server) -> CommsResult<()> {
-        todo!()
+        match self.connections.get(addr) {
+            None => {
+                self.connections
+                    .insert(addr.to_string(), Connection::Open(server));
+                match self.send(
+                    addr,
+                    Message {
+                        msg_type: MessageType::Handshake,
+                        load: self.ip.clone(),
+                    },
+                ) {
+                    Ok(_) => Ok(()),
+                    Err(e) => Err(e),
+                }
+            }
+            Some(_) => Err(CommsError::ConnectionExists(addr.to_string())),
+        }
     }
 
     // Sends the provided message to the server at the given `addr`.
     // Can only send messages through open connections. If the server
     // responds with a ServerLimitReached error, its corresponding connection
     // should be closed.
     fn send(&mut self, addr: &str, msg: Message) -> CommsResult<Response> {
-        // server.receive(msg)
-        todo!()
+        match self.connections.get_mut(addr) {
+            None => Err(CommsError::ConnectionNotFound(addr.to_string())),
+            Some(conn) => match conn {
+                Connection::Closed => Err(CommsError::ConnectionClosed(addr.to_string())),
+                Connection::Open(server) => {
+                    let result = server.receive(msg);
+                    if let Err(CommsError::ServerLimitReached(_)) = result {
+                        *self.connections.get_mut(addr).unwrap() = Connection::Closed;
+                    }
+                    result
+                }
+            },
+        }
     }
 
     // Returns whether the connection to `addr` exists and has
     // the `Open` status.
     #[allow(dead_code)]
     fn is_open(&self, addr: &str) -> bool {
-        todo!()
+        match self.connections.get(addr).unwrap() {
```

Comment:
W taki sposób `is_open()` panikuje, gdy zostanie wywołane dla klucza, którego nie ma w mapie - nie jest to oczekiwane zachowanie